lib/ecto/adapters/sql.ex

```diff
@@ -141,11 +146,17 @@ defmodule Ecto.Adapters.SQL do
   def to_sql(kind, repo, queryable) do
     adapter = repo.__adapter__
 
-    {_meta, prepared, params} =
+    plan =
       Ecto.Queryable.to_query(queryable)
       |> Ecto.Query.Planner.query(kind, repo, adapter)
-
-    {prepared, params}
+    case plan do
+      {:execute, _meta, %{statement: statement}, params} ->
```

Comment:
Another option is to remove this knowledge from Ecto and pass it to the adapter completely.

1. The call to Ecto.Query.Planner would return a key indicating `:nocache` (execute_planned),  `:cache` (prepare_execute_planned), `:cached` (execute_prepared).

2. This would be sent to the adapter execute which would act accordingly. This way the adapter is a single function, decoupled from SQL semantics, and you can do as you wish from the adapter side. The only difference is that in the prepare_execute case, the adapter would have to call the anonymous function as we won't do it in the repository.